core/src/main/c/share/net.c

```diff
@@ -43,17 +43,31 @@ int set_int_sockopt(int fd, int level, int opt, int value) {
 JNIEXPORT jint JNICALL Java_io_questdb_network_Net_socketTcp0
         (JNIEnv *e, jclass cl, jboolean blocking) {
     int fd = socket(AF_INET, SOCK_STREAM, 0);
-    if (fd > 0 && !blocking) {
-        if (fcntl(fd, F_SETFL, O_NONBLOCK) < 0) {
-            close(fd);
-            return -1;
+    if (fd > 0) {
+        if (!blocking) {
+            if (fcntl(fd, F_SETFL, O_NONBLOCK) < 0) {
+                close(fd);
+                return -1;
+            }
+
+            int oni = 1;
+            if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (char *) &oni, sizeof(oni)) < 0) {
+                close(fd);
+                return -1;
+            }
         }
 
-        int oni = 1;
-        if (setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, (char *) &oni, sizeof(oni)) < 0) {
-            close(fd);
-            return -1;
-        }
+        // enable TCP keepalive probes, otherwise firewalls and load balancers may drop connections
+        // during a period of inactivity
+        set_int_sockopt(fd, SOL_SOCKET, SO_KEEPALIVE, 1);
```

Comment:
nit: we should check errors here and below. Otherwise, we won't notice if these calls do nothing instead of the intended action.

Reply:
I was considering that. but what do you if the options cannot be set? fail questdb? I thought best-effort is OK in this case. wdyt?

Reply:
Well, at least we could log something. But we could even fail the socket creation as a whole - that's to detect compatibility issues (very unlikely to happen).

Reply:
I was considering creating a separate private native function `Net.setKeepAlive()` and call it from `socketTcp()` and guard it with a system property. Then it would make sense to check return values - because if on some OS the options cannon be set then we could use the system property as a fallback and disable it. But when it's done unconditionally on all sockets then I don't dare to fail socket creation just because KeepAlive cannot be set.

Reply:
TBH I really like your initial approach, i.e. a separate method. I don't think an extra JNI call overhead matters when it comes to socket creation. WDYT?

Reply:
I was not worried about JNI overhead, I thought it was unnecessarily complex. but I can do it like if you find it better.

Reply:
To me, we would be able at least to log errors with a standalone method. I don't like the idea of swallowing errors in the native code.